File: plugins/de-groot/models/category.py
from database.db import db
from flask_mongoengine import BaseQuerySet
import datetime
from slugify import slugify, Slugify, UniqueSlugify


class Category(db.Document):
    name = db.StringField(required=True)
    sort = db.IntField(required=False, default=0)
    countries = db.ListField(required=False, default=[])
    sku = db.StringField(required=False, default='')
    system_key = db.StringField(required=True, unique=True)
    display_name = db.ListField(required=True, default=[])
    slug = db.StringField(required=True, unique=True)
    description = db.StringField(required=False, default=None)
    shareable = db.BooleanField(required=False, default=False)
    published = db.BooleanField(required=False, default=True)
    media = db.ListField(required=False)
    price_build = db.DictField(required=False, default={})
    has_products = db.BooleanField(required=False, default=False)
    has_manifest = db.BooleanField(required=False, default=False)
    calculation_method = db.ListField(required=False, )
    dlv_days = db.ListField(required=False, default=[])
    printing_method = db.ListField(required=False, default=None)
    production_days = db.ListField(required=False, default=[])
    production_dlv = db.ListField(required=False, default=[])
    free_entry = db.ListField(required=False, default=[])
    limits = db.ListField(required=False, default=[])
    ranges = db.ListField(required=False, default=[])
    range_list = db.ListField(required=False, default=[])
    ref_id = db.StringField(required=False, default='')
    ref_category_name = db.StringField(required=False, default='')
    start_cost = db.IntField(required=False, default=0)
    vat = db.IntField(required=False, default=0)
    bleed = db.IntField(required=False, default=0)
    range_around = db.IntField(required=False, default=0)
    created_at = db.DateTimeField(default=datetime.datetime.now, required=True)
    checked = db.BooleanField(default=False)
    additional = db.DictField(required=False, default={})

    meta = {
        'indexes': ['-created_at', 'slug'],
        'ordering': ['-created_at', 'name'],
        'collection': 'categories',
        'queryset_class': BaseQuerySet
    }

    def __setattr__(self, key, value):
        super(Category, self).__setattr__(key, value)
        if key == 'name':
            if value is not None:
                self.slug = slugify(value, to_lower=True)

File: microservices/categories/categories/Models/Category.js
const mongoose = require('mongoose');
const slugify = require('slugify');

const categorySchema = new mongoose.Schema({
    name: { type: String, required: true },
    sort: { type: Number, required: false, default: 0 },
    countries: { type: [String], required: false, default: [] },
    sku: { type: String, required: false, default: '' },
    system_key: { type: String, required: true, unique: true },
    display_name: { type: [Object], required: true, default: [] },
    slug: { type: String, required: true, unique: true },
    description: { type: String, required: false, default: null },
    shareable: { type: Boolean, required: false, default: false },
    published: { type: Boolean, required: false, default: true },
    media: { type: [Object], required: false, default: [] },
    price_build: { type: Object, required: false, default: {} },
    has_products: { type: Boolean, required: false, default: false },
    has_manifest: { type: Boolean, required: false, default: false },
    calculation_method: { type: [String], required: false, default: [] },
    dlv_days: { type: [String], required: false, default: [] },
    printing_method: { type: [String], required: false, default: [] },
    production_days: { type: [String], required: false, default: [] },
    production_dlv: { type: [String], required: false, default: [] },
    free_entry: { type: [String], required: false, default: [] },
    limits: { type: [Object], required: false, default: [] },
    ranges: { type: [Object], required: false, default: [] },
    range_list: { type: [Object], required: false, default: [] },
    ref_id: { type: String, required: false, default: '' },
    ref_category_name: { type: String, required: false, default: '' },
    start_cost: { type: Number, required: false, default: 0 },
    vat: { type: Number, required: false, default: 0 },
    bleed: { type: Number, required: false, default: 0 },
    range_around: { type: Number, required: false, default: 0 },
    created_at: { type: Date, default: Date.now, required: true },
    checked: { type: Boolean, default: false },
    additional: { type: Object, required: false, default: {} }
}, {
    versionKey: false
});

categorySchema.pre('save', function (next) {
    if (this.name) {
        this.slug = slugify(this.name, { lower: true });
        this.system_key = slugify(this.name, { lower: true });
    }
    next();
});

const Category = mongoose.model('Category', categorySchema);

module.exports = Category;